### ISSUE TITLE
Token '1614408909' doesn't match specifier '%x'

### ISSUE THREAD

**Issue by luawtf** [allinurl/goaccess#2045]
Hello, I'm trying to use GoAccess v1.4.5 to parse log files generated by Caddy v2.3.0 (latest) using the default `CADDY JSON Structured` log format:
```json
{"ts":"%x.%^","request":{"remote_addr":"%h:%^","proto":"%H""method":"%m","host":"%v","uri":"%U","headers":{"User-Agent":["%u"]},"tls":{"cipher_suite":"%k","proto":"%K"}},"duration":"%T","size":"%b","status":"%s","resp_headers":{"Content-Type":["%M"]}}
```
But I'm receiving errors like `Token '1614408909' doesn't match specifier '%x'` for log messages like:
```json
{"level":"info","ts":1614408909.3341959,"logger":"http.log.access.log0","msg":"handled request","request":{"remote_addr":"50.68.124.60:46556","proto":"HTTP/2.0","method":"GET","host":"ipfs.foxgirl.dev","uri":"/QmcniBv7UQ4gGPQQW2BwbD4ZZHzN3o3tPuNLZCbBchd1zh","headers":{"Range":["bytes=24903680-"],"Te":["trailers"],"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8"],"Accept-Encoding":["gzip, deflate, br"],"Dnt":["1"],"Upgrade-Insecure-Requests":["1"],"User-Agent":["Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:85.0) Gecko/20100101 Firefox/85.0"],"Accept-Language":["en-CA,en-US;q=0.7,en;q=0.3"],"Referer":["https://t.co/WV8GcwDZVy?amp=1"],"If-Range":["\"QmcniBv7UQ4gGPQQW2BwbD4ZZHzN3o3tPuNLZCbBchd1zh\""]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"h2","proto_mutual":true,"server_name":"ipfs.foxgirl.dev"}},"common_log":"50.68.124.60 - - [27/Feb/2021:06:55:09 +0000] \"GET /QmcniBv7UQ4gGPQQW2BwbD4ZZHzN3o3tPuNLZCbBchd1zh HTTP/2.0\" 206 14483456","duration":408.211730412,"size":14483456,"status":206,"resp_headers":{"Date":["Sat, 27 Feb 2021 06:48:21 GMT"],"Content-Length":["22399601"],"Last-Modified":["Thu, 01 Jan 1970 00:00:01 GMT"],"Server":["Caddy"],"Access-Control-Allow-Methods":["GET"],"Content-Type":["video/mp4"],"X-Ipfs-Path":["/ipfs/QmcniBv7UQ4gGPQQW2BwbD4ZZHzN3o3tPuNLZCbBchd1zh"],"Access-Control-Expose-Headers":["Content-Range","X-Chunked-Output","X-Stream-Output"],"Accept-Ranges":["bytes"],"Cache-Control":["public, max-age=29030400, immutable"],"Access-Control-Allow-Headers":["*","Content-Type","Range","User-Agent","X-Requested-With"],"Access-Control-Allow-Origin":["ipfs.foxgirl.dev"],"Etag":["\"QmcniBv7UQ4gGPQQW2BwbD4ZZHzN3o3tPuNLZCbBchd1zh\""],"Content-Range":["bytes 24903680-47303280/47303281"]}}
```
This log message was generated with the `log` directive:
```caddyfile
log {
  output file /logs/ipfs--foxgirl--dev.log {
    roll_size 1gb
    roll_keep 10
    roll_keep_for 2160h
  }
  format json
  level DEBUG
}
```
Any ideas? Thank you for your time!

**Comment by tysg** [comment]
Are you configuring the `log-format` in `goaccess.conf`? Seems like the default config line for Caddy is outdated. It should be:

```
{ "ts": "%x.%^", "request": { "remote_addr": "%h:%^", "proto":"%H", "method": "%m", "host": "%v", "uri": "%U", "headers": {"User-Agent": ["%u"], "Referer": ["%R"] }, "tls": { "cipher_suite":"%k", "proto": "%K" } }, "duration": "%T", "size": "%b","status": "%s", "resp_headers": { "Content-Type": ["%M"] } }
```

**Comment by luawtf** [comment]
I am using a fresh install of GoAccess with no configuration changes, running on an almost-fresh install of Alpine Linux edge.

**Comment by luawtf** [comment]
I have an idea of what may be causing the issue! Take a look at the code for parsing `%x` tokens:
https://github.com/allinurl/goaccess/blob/057d65683cdc3e25d943abf264375ec6d2021a9c/src/parser.c#L1233-L1249
I'm pretty sure the problem lies in the call to `str_to_time`, and I'll explain why. Take a look at `str_to_time`'s code:
https://github.com/allinurl/goaccess/blob/057d65683cdc3e25d943abf264375ec6d2021a9c/src/util.c#L496-L541
See all those comments / preprocessor conditions about having glibc? Well, Alpine linux doesn't have glibc! It uses the musl C stdlib instead. Could that be the problem?

**Comment by luawtf** [comment]
I have pretty much confirmed that it is a problem with the lack of glibc, see this video for a demonstration:
[https://youtu.be/rCyjoA7jUH0](https://youtu.be/rCyjoA7jUH0)
TL;DR: GoAccess fails to parse timestamps correctly on systems using stdlibs other than glibc.
Both of these `goaccess` binaries are the exact same version of GoAccess (v1.4.5 tarball from [https://tar.goaccess.io/goaccess-1.4.5.tar.gz](https://tar.goaccess.io/goaccess-1.4.5.tar.gz)) compiled with the same options and the same toolchain (gcc, GNU make).

**Comment by allinurl** [comment]
Thanks a lot for reporting this and sharing those details. I believe this has been addressed upstream on [**this commit**](https://github.com/allinurl/goaccess/commit/cc0545363210a090c4b3d03e953ddd503f88c96f). Can I ask you to please build from [**development**](https://github.com/allinurl/goaccess#build-from-github-development) and let me know if that fixes the issue on your end?

**Comment by luawtf** [comment]
Confirmed fixed when building master, thanks!

**Comment by allinurl** [comment]
Awesome, thanks for the update. The next release will be pushed out soon.
